doc/HORDES.md

```diff
@@ -0,0 +1,62 @@
+Needs O(1) lookup for next group to process (will likely iterate on it).
+O(1) insertion into a new monster_group iff next update for it is sooner than the next update for the monster being inserted (likely, it just took a turn), O(logn) insertion if new monster group updates later than the new monster, or does not exist, requiring an insertion into monster_turn_sequence.
+O(logn) re-insertion into the current monster_group, since its turn count is definitely changing.
+Optimizations: Batch insertions into neighbor monster groups to minimize number of insertions of the group.  Process monsters slightly ahead of schedule (based on distance from player?) to reduce number of updates.
+
+Reacting to stimuli: (sound or visible stimuli such as flashing lightsor movement)
+Iterate over monster_map locations in range of the event (O(n^2) in range) and check for replacing existing event (loudest sound or most visible stimuli wins).  Weird special case, underground gets extra extra attenuation.  At a minimum queue these up and process them once per turn, with only loudest sound per submap (or possibly loudest per map? being applied to all monster groups.
+For flavor it would be nice if there were some random-ish noises on the overmap as well.
+
+For map data (navigability, vision, and scent), monster AI can have a persistent "oracle" injected into it, which it can query for map data such as nearest visible enemy, nearest strong scent, and pathing information.  This means the monster AI doesn't have a dependency on the overmap, and creates an injection point for testing monster AI.
+
+Save/load:
+When an overmap is savesd, the monster groups can be written out in no particular order.
+When an overmap is loaded, re-inserting the monster groups should repopulate the data structures properly.
+
+API on overmap (or overmapbuffer?):
+despawn_monsters() // Move monsters from map to an overmap.
+spawn_monsters()
+move_hordes()
+add_monster() // Will probably want it for missions and mapgen
+
+API on monster_map
+add_monster(monster new_monster)
+add_monster_group(monster_group new_monsters)
+save_monsters()
+load_monsters()?
+monster_group_info get_monster_group_info(tripoint location) // For overmap display
+move()
+
+API on monster_group
+add_monster(monster new_monster)
+merge_monster_group(monster_group new_monsters)
+
+
+
+Movement:
+Each OMT has a load factor determined as a low estimate of the number of navigable tiles for a tile of its kind.
+OMTs backing generated submaps should be updated with the actual load factor.
+Monsters attempting to move into a submap at capacity should lose a turn.
+OMTs should have a move cost derived in a similar way to the load factor.
+Monster pays move cost to exit an OMT
+Eventually monster groups should fight each other if they encounter opposition.
+Dense hordes should mark map tiles for destruction as they cross over them, at some point becoming completely wrecked.
```

Comment:
Many players have found that it's currently a viable strategy to stay very quiet in your house and wait for the horde to pass. They will rarely mess anything up. So having them behave differently on overmap scale would be strange.